core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala

```diff
@@ -264,16 +239,95 @@ class SimpleAclAuthorizer extends Authorizer with Logging {
     authorizerLogger.debug(s"Principal = $principal is $permissionType Operation = $operation from host = $host on resource = $resource")
   }
 
+  /**
+    * Safely updates the resources acls by ensureing reads and writes respect the expected zookeeper version.
+    * Continues to retry until it succesfully updates.
+    *
+    * @param resource the resource to change acls for
+    * @param getNewAcls function to transform existing acls to new acls
+    * @return boolean indicating if a change was made
+    */
+  private def updateResourceAcls(resource: Resource)(getNewAcls: Set[Acl] => Set[Acl]): Boolean = {
+    val path = toResourcePath(resource)
+
+    var currentVersionedAcls =
+      if(aclCache.contains(resource))
+        getAclsFromCache(resource)
+      else
+      getAclsFromZk(resource)
+    var newVersionedAcls = currentVersionedAcls
+    var writeComplete = false
+    while (!writeComplete) {
+      val newAcls = getNewAcls(currentVersionedAcls.acls)
+      val data = Json.encode(Acl.toJsonCompatibleMap(newAcls))
+      try {
+        val (updateSucceeded, updateVersion) = zkUtils.conditionalUpdatePersistentPathIfExists(path, data, currentVersionedAcls.version)
+        if(!updateSucceeded) {
+          trace(s"Failed to update acls for $resource. Used version ${currentVersionedAcls.version}. Reading data and retrying update.")
+          currentVersionedAcls = getAclsFromZk(resource);
+        } else {
+          if(newAcls.isEmpty) {
+            trace(s"Deleting path for $resource because it had no acls remaining")
+            zkUtils.deletePath(path) // It would be safest to use the versioned request, but this zkClient doesn't support it
+          }
+          newVersionedAcls = VersionedAcls(newAcls, updateVersion)
+        }
+        writeComplete = updateSucceeded
+      } catch {
+        case e: ZkNoNodeException =>
+          try {
+            debug(s"Node for $resource does not exist, attempting to create it.")
+            zkUtils.createPersistentPath(path, data)
```

Comment:
This is going to be a style comment, so it is arguable whether it is better or not. I'd rather have this try/catch wrapping the call to `conditionalUpdatePersistentPathIfExists` because we are specifically interested in the exception that it can throw. I think it makes the intent more clear if we move it there, but again, up to you.